### PR TITLE
Setup loss_type in config at model init time

### LINKED ISSUE #34615: Torch.compile Graph break introduced due to new loss function api
### System Info

- `transformers` version: 4.47.0.dev0
- Platform: Linux-5.14.0-284.73.1.el9_2.x86_64-x86_64-with-glibc2.31
- Python version: 3.10.15
- Huggingface_hub version: 0.26.2
- Safetensors version: 0.4.5
- Accelerate version: 1.0.1
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.4.1+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: Yes
- GPU type: NVIDIA A100-SXM4-80GB

### Who can help?

@ArthurZucker

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

PR https://github.com/huggingface/transformers/pull/34191 introduces a new Loss API. Post this PR, Dynamo was broken, which was identified and fixed in this issue: https://github.com/huggingface/transformers/issues/34402. Post this (on master), Dynamo runs without errors.

However, in this process, a new Graph Break has been introduced due to [this line](https://github.com/huggingface/transformers/pull/34191/files#diff-6b72b98c4c2dcfc6cc606843917733f5d858374fbc22a735ff483bbc0c1e63eaR4992):
```
loss_type = re.findall(loss_groups, self.__class__.__name__)
```
This is due to the new regex check.

Since the dispatch function actually checks for an attr on the config, the fix for this is quite simple - set the loss_type at model init time itself.

### Expected behavior

No additional graph breaks.

### PR DESCRIPTION
**Pull request by ChanderG**
ensures no additional graph break introduced when torch.compile'ed

fixes #34615

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by ChanderG** [comment]
If this approach is fine - I will extend the PR to cover a wide set of models.

**Review by qubvel — COMMENTED** [review]
Hi @ChanderG! Thanks for opening an issue and PR! Nice catch with the graph breaks!
Re the PR it is usually not the best decision to modify class attribute inplace, however this pattern is widely used across classification models losses
```
if self.num_labels == 1:
    self.config.problem_type = "regression"
elif self.num_labels > 1 and (labels.dtype == torch.long or labels.dtype == torch.int):
    self.config.problem_type = "single_label_classification"
else:
    self.config.problem_type = "multi_label_classification"
```
do you have other ideas on how to handle it without modifying the config attribute? Is there a way to fix `re.findall ` instead of introducing the change for all models?

Alternatively we can have a model attribute, smth like
```python
self.loss_type = "ForCausalLM" if self.config.loss_type is None else self.config.loss_type
```
but not feel strong re this :)

**Comment by ChanderG** [comment]
A more generic option may be to run the regex call at model init time, instead of hardcoding. I need to check, but it may still be efficient placed there (ie, won't cause a graph break from init).

**Review by ArthurZucker — COMMENTED** [review]
Hey! We should rather do this in the call of `super().__init__(config)` as otherwise we have to monkey patch each and every model!

**Comment by ChanderG** [comment]
@ArthurZucker Updated to generic lookup. Tested with the repro here https://github.com/huggingface/transformers/issues/34615#issuecomment-2482082014 and no Graph Breaks with this latest version as well.

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1437,6 +1437,16 @@ def __init__(self, config: PretrainedConfig, *inputs, **kwargs):
...
+        if getattr(config, "loss_type", None) is None:
+            loss_type = self.__class__.__name__
+            if loss_type not in LOSS_MAPPING:
+                loss_groups = f"({'|'.join(LOSS_MAPPING)})"
+                loss_type = re.findall(loss_groups, self.__class__.__name__)
+                if len(loss_type) > 0:
+                    loss_type = loss_type[0]
+                else:
+                    loss_type = None
+            self.config.loss_type = loss_type
```
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:1449]: As 
```
            loss_type = self.__class__.__name__
            if loss_type not in LOSS_MAPPING:
                loss_groups = f"({'|'.join(LOSS_MAPPING)})"
                loss_type = re.findall(loss_groups, self.__class__.__name__)
                if len(loss_type) > 0:
                    loss_type = loss_type[0]
                else:
                    loss_type = None
```

is used twice we can probably have a function for `find_class_loss_function`and keep a single source of truth! 🤗 

https://github.com/huggingface/transformers/blob/00962d103459218c6f353da319a1d12adcd56a15/src/transformers/modeling_utils.py#L5044-L5051

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (ChanderG) [reply src/transformers/modeling_utils.py:1449]: I was thinking that if we are going to set the loss_type into the config at init time, we can remove the entire check from the loss_function method and simplify it to the following:
```
    def loss_function(self):
        loss_type = getattr(self.config, "loss_type", None)

        if loss_type is None or loss_type not in LOSS_MAPPING:
            logger.warning_once(
                f"`loss_type={loss_type}` was set in the config but it is unrecognised."
                f"Using the default loss: `ForCausalLMLoss`."
            )
            loss_type = "ForCausalLM"
        return LOSS_MAPPING[loss_type]
```
What do you think?

Comment (ArthurZucker) [reply src/transformers/modeling_utils.py:1449]: yep sounds good!

**Comment by ArthurZucker** [comment]
Waiting on this to put in the upcoming release !! 🤗

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (ChanderG) [reply src/transformers/modeling_utils.py:1449]: Done!

**Comment by ArthurZucker** [comment]
Okay this is failing tests because the `loss_type` is stored in the `config`, which is not reset in the tests. 
It is actually be better to store the `loss_type` in the `PreTrainedModel` rather than in the config!

**Comment by ArthurZucker** [comment]
Pushed directly, hope it's not a problem 🤗 wanted to get this fix in fast!

**Comment by ArthurZucker** [comment]
Thanks for your patience! I thought this had been merged a while ago!
